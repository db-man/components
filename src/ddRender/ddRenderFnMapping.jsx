import React from 'react';
/**
 * https://github.com/handlebars-lang/handlebars.js/issues/1174
 * ```
 * Compiled with problems:
 * ERROR in ./node_modules/handlebars/lib/index.js 15:11-24
 * Module not found: Error: Can't resolve 'fs' in '/.../node_modules/handlebars/lib'
 * ```
 */
// import Handlebars from "handlebars";
import Handlebars from 'handlebars/dist/handlebars';

import {
  ImageLink, ImageLinks, Link, Links, Fragment,
} from 'components/Links';
import ErrorAlert from 'components/ErrorAlert';

/**
 * tpl: {{#replace "foo" "bar"}}{{title}}{{/replace}}
 * input: {title:"foo"}
 * output: bar
 */
Handlebars.registerHelper('replace', function replaceHelper(find, replace, options) {
  const string = options.fn(this);
  return string.replace(find, replace);
});

/**
 * tpl: {{join record.tags ", "}}
 * input: {"tags": ["foo", "bar"]}
 * output: "foo, bar"
 */
Handlebars.registerHelper('join', (arr, sep) => {
  if (!arr) return '';
  return arr.join(sep);
});

const ddComponent = (Component) => function dComponent(val, record, index, args /* , col */) {
  if (!args || !args[1]) {
    return <Component>{val}</Component>;
  }

  const tplStr = args[1];
  const tpl = Handlebars.compile(tplStr);
  const json = tpl({ record });
  try {
    // {foo:'bar'} <= "{\"foo\":\"bar\"}"
    // "foo" <= "\"foo\""
    const props = JSON.parse(json);
    if (typeof props === 'string') {
      return <Component>{props}</Component>;
    }
    return <Component {...props} />; // eslint-disable-line react/jsx-props-no-spreading
  } catch (err) {
    console.error('Failed to parse JSON for tpl, err:', err, json); // eslint-disable-line no-console
    return (
      <div>
        val:
        {' '}
        {val}
        <ErrorAlert json={json} error={err} tplStr={tplStr} record={record} />
      </div>
    );
  }
};

/**
 * Data Driving Render Function Mapping
 */
const ddRenderFnMapping = {
  // renderFnName: (cellVal, rowVal, colIndex, ...renderFnStr) => null
  // renderFnName: (cellVal, rowVal, colIndex, ...renderFnProps) => null

  /**
   * @param {number|string} val e.g. "docs"
   * @param {Object} record e.g. `{"org":"github","repo":"docs"}`
   * @param {number} index e.g. 9, it's the 10th row
   * @param {Object|undefined} args e.g `["Link", "{\"href\":\"https://github.com/{{record.org}}/{{record.repo}}\",\"text\":\"{{record.repo}}\"}"]`
   *                                args could be undefined when column is {"type:listPage": "Link"}
   *  Usage:
   * 1. "type:getPage": ["Link", "{\"href\":\"https://github.com/{{record.org}}/{{record.repo}}\",\"text\":\"{{record.repo}}\"}"],
   * 2. "type:listPage": ["Link", "{\"href\":\"https://github.com/{{record.org}}/{{record.repo}}\",\"text\":\"{{record.repo}}\"}"]
   * 3. "type:listPage": "Link"
   */
  Link: ddComponent(Link),
  /**
   * Usage:
   * ```json
   * {"type:listPage": ["ImageLink",
   *   "{\"url\":\"{{record.url}}\",\"imgSrc\":\"{{record.url}}_th.jpg\",
   *     \"tags\":\"{{record.tags}}\"}"
   * ]}
   * ```
   * Issues:
   *   When record.url="https://foo.com/get?foo=bar".
   *   By using `{{record.url}}` will generate "https://foo.com/get?foo&#x3D;bar".
   *   To prevent this issue, `{{{record.url}}}` could be used. (https://handlebarsjs.com/guide/expressions.html#html-escaping)
   */
  ImageLink: ddComponent(ImageLink),
  ImageLinks: ddComponent(ImageLinks),
  /**
   * Usage:
   * ```json
   * {"type:listPage": ["Fragment", "{{join record.tags \", \"}}"]}
   * ```
   */
  Fragment: ddComponent(Fragment),
  Links: ddComponent(Links),
};

export default ddRenderFnMapping;
