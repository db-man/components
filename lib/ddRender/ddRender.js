import * as constants from '../constants';
import ddRenderFnMapping from './ddRenderFnMapping'; // Default render func when "type:listPage" or "type:getPage" not defined in db table column

const defaultRenders = {
  [constants.NUMBER]: val => val,
  [constants.STRING]: val => val,
  [constants.STRING_ARRAY]: val => val && val.join(', '),
  [constants.BOOL]: val => val === undefined ? '' : String(val)
};
/**
 * @param {string|string[]} args
 * @param {Object} tplExtra
 */

export const getRender = (args, tplExtra) => {
  // the column render function defined in Table component of antd
  // renderFn = (val, record, index) => ()
  let renderFn;

  if (!args) {
    return renderFn;
  }

  if (typeof args === 'string') {
    const fn = ddRenderFnMapping[args];

    if (fn) {
      renderFn = fn;
    }
  }

  if (Array.isArray(args)) {
    const [renderFnName] = args;

    renderFn = (val, record, index) => ddRenderFnMapping[renderFnName](val, record, index, args, tplExtra);
  }

  return renderFn;
};
/**
 * column def:
 * {
 *   "type:listPage": ["Link", "{{record.url}}"]
 * }
 */

export const getColumnRender = column => {
  const customRender = getRender(column['type:listPage']);

  if (customRender) {
    return customRender;
  } // the column render function defined in Table component of antd
  // renderFn = (val, record, index) => ()


  return defaultRenders[column.type || constants.STRING];
};
export const getDetailPageColumnRender = (column, tplExtra) => {
  const customRender = getRender(column['type:getPage'], tplExtra);

  if (customRender) {
    return customRender;
  } // the column render function defined in Table component of antd
  // renderFn = (val, record, index) => ()


  return defaultRenders[column.type || constants.STRING];
};
export const getRenderResultByColumn = (value, record, index, args, column) => getColumnRender(column)(value, record, index); // eslint-disable-line max-len