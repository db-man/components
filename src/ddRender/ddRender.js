import * as constants from 'constants.js';
import ddRenderFnMapping from './ddRenderFnMapping';

export const getColumnRender = (column) => {
  // the column render function defined in Table component of antd
  // renderFn = (val, record, index) => ()
  let renderFn;

  if (column.type === constants.STRING_ARRAY) {
    // Default render func when "type:listPage" not defined in table column
    renderFn = (val) => val && val.join(', ');
  }

  if (column['type:listPage']) {
    if (typeof column['type:listPage'] === 'string') {
      const fn = ddRenderFnMapping[column['type:listPage']];
      if (fn) {
        renderFn = fn;
      }
    }
    /**
     * column def:
     * {
     *   "type:listPage": ["Link", "{{record.url}}"]
     * }
     */
    if (Array.isArray(column['type:listPage'])) {
      const [renderFnName] = column['type:listPage'];
      renderFn = (val, record, index) => ddRenderFnMapping[renderFnName](
        val,
        record,
        index,
        column['type:listPage'],
        column,
      );
    }
  }

  return renderFn;
};

export const getRenderResultByColumn = (value, record, index, args, column) => getColumnRender(column)(value, record, index); // eslint-disable-line max-len
