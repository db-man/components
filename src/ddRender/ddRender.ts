import * as constants from '../constants';
import DbColumn, { RenderKeyType } from '../types/DbColumn';
import { RowType } from '../types/Data';
import { RenderArgs } from '../types/UiType';
import ddRenderFnMapping from './ddRenderFnMapping';

// Default render func when "type:listPage" or "type:getPage" not defined in db table column
const defaultRenders = {
  [constants.NUMBER]: (val: number) => val,
  [constants.STRING]: (val: string) => val,
  [constants.STRING_ARRAY]: (val: string[]) => val && val.join(', '),
  [constants.BOOL]: (val: boolean) => (val === undefined ? '' : String(val)),
};

/**
 * @param {string|string[]} args e.g. "Link" or ["Link", "{{record.url}}"]
 * @param {Object} tplExtra
 */
export const getRender = (args: RenderArgs, tplExtra?: any) => {
  // the column render function defined in Table component of antd
  // renderFn = (val, record, index) => ()
  let renderFn;

  if (!args) {
    return renderFn;
  }

  if (typeof args === 'string') {
    const fn = ddRenderFnMapping[args];
    if (fn) {
      renderFn = fn;
    }
  }

  if (Array.isArray(args)) {
    const [renderFnName] = args;
    renderFn = (val: any, record: RowType, index?: number) =>
      ddRenderFnMapping[renderFnName](val, record, index, args, tplExtra);
  }

  return renderFn;
};

/**
 * column def:
 * {
 *   "type:listPage": ["Link", "{{record.url}}"]
 * }
 */
export const getColumnRender = (
  renderKey: RenderKeyType,
  column: DbColumn,
  tplExtra?: any
) => {
  const customRender = getRender(column[renderKey], tplExtra);
  if (customRender) {
    return customRender;
  }

  // the column render function defined in Table component of antd
  // renderFn = (val, record, index) => ()
  return defaultRenders[column.type || constants.STRING];
};

// export const getRenderResultByColumn = (
//   value: any,
//   record: RowType,
//   index: number,
//   args: RenderArgs,
//   column: Column
// ) => getColumnRender('type:listPage', column)(value, record, index);
