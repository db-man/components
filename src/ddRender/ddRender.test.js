import { getColumnRender } from './ddRender';

describe('getColumnRender', () => {
  test('should return proper value', () => {
    const view = getColumnRender({
      id: 'url',
      'type:listPage': 'ImageLink',
    });
    expect(view('https://a.com/b.jpg', { url: 'https://a.com/b.jpg' }, 1)).toMatchSnapshot();
  });

  test('should return proper value2', () => {
    const view = getColumnRender({
      id: 'yrl',
      'type:listPage': ['ImageLink', '{"url":"{{record.url}}","imgSrc":"{{record.url}}"}'],
    });
    expect(view('https://a.com/large/b.jpg', { url: 'https://a.com/large/b.jpg' })).toMatchSnapshot();
  });

  describe('given hidden column in a list page', () => {
    it('should return undefined render func', () => {
      const colFunc = getColumnRender({ 'type:listPage': 'HIDE' });
      expect(colFunc).toBe(undefined);
    });
  });
});

// describe('getRenderResultByColumn', () => {
//   test('should return proper value', () => {
//     const el = getRenderResultByColumn(
//       'foo',
//       { name: 'foo' },
//       0,
//       'HIDE',
//       { 'type:listPage': 'HIDE' },
//     );
//     expect(el).toMatchSnapshot();
//   });
// });
