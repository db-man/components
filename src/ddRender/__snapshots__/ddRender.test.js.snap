// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getColumnRender given Link should match snapshot 1`] = `
<Link
  href="https://foo.com"
  text="https://foo.com"
>
  
</Link>
`;

exports[`getColumnRender should return proper value 1`] = `
<ImageLink
  description=""
  imgSrc=""
  url=""
>
  https://a.com/b.jpg
</ImageLink>
`;

exports[`getColumnRender should return proper value2 1`] = `
<ImageLink
  description=""
  imgSrc="https://a.com/large/b.jpg"
  url="https://a.com/large/b.jpg"
>
  
</ImageLink>
`;

exports[`getRender given simple arg should render properly 1`] = `
<Link
  href=""
  text=""
>
  https://foo.com
</Link>
`;

exports[`getRender given tpl should render properly 1`] = `
<Link
  href="https://foo.com"
  text="https://foo.com"
>
  
</Link>
`;
